python/ray/tune/execution/ray_trial_executor.py

```diff
@@ -732,37 +817,75 @@ def has_resources_for_trial(self, trial: Trial) -> bool:
             boolean
 
         """
+        resource_request = trial.placement_group_factory
+
         return (
             trial in self._staged_trials
-            or (
-                len(self._cached_actor_pg) > 0
-                and (self._pg_manager.has_cached_pg(trial.placement_group_factory))
-            )
-            or self._pg_manager.can_stage()
-            or self._pg_manager.has_ready(trial, update=True)
-            or self._pg_manager.has_staging(trial)
+            or self._cached_resources_to_actor[resource_request]
+            or len(self._staged_trials) < self._max_staged_actors
+            or self._resource_manager.has_resources_ready(resource_request)
         )
 
+    def _occupied_resources(self) -> dict:
+        total_resources = {"CPU": 0, "GPU": 0}
+        for allocated_resource in self._trial_to_allocated_resources.values():
+            resource_request = allocated_resource.resource_request
+            for bundle_resources in resource_request.bundles:
+                for key, val in bundle_resources.items():
+                    total_resources[key] = total_resources.get(key, 0) + val
+        return total_resources
+
     def debug_string(self) -> str:
         """Returns a human readable message for printing to the console."""
-        total_resources = self._pg_manager.occupied_resources()
-        return self._resource_updater.debug_string(total_resources)
+        occupied_resources = self._occupied_resources()
+
+        return self._resource_updater.debug_string(occupied_resources)
 
-    def on_step_begin(self, trials: List[Trial]) -> None:
+    def on_step_begin(self) -> None:
         """Before step() is called, update the available resources."""
         self._resource_updater.update_avail_resources()
-        self._trial_just_finished_before = self._trial_just_finished
-        self._trial_just_finished = False
 
-    def on_step_end(self, trials: List[Trial]) -> None:
+    def on_step_end(self) -> None:
+        self._cleanup_cached_actors()
         self._do_force_trial_cleanup()
-        if time.time() > self.last_pg_recon + self.pg_recon_interval:
-            # Only do this every now and then - usually the placement groups
-            # should not get out of sync, and calling this often is inefficient
-            self._pg_manager.reconcile_placement_groups(trials)
-            self.last_pg_recon = time.time()
 
-        self._pg_manager.cleanup()
+    def _count_staged_resources(self):
+        counter = Counter()
+        for trial in self._staged_trials:
+            resource_request = trial.placement_group_factory
+            counter[resource_request] += 1
+        return counter
+
+    def _cleanup_cached_actors(self, force_all: bool = False):
```

Comment:
we should document why do we need this?
Or why is this "cron job" necessary for keeping consistency?

Reply:
I think there's two things here

1) (Gracefully) clean up cached actors (e.g. on interrupt) - we need this anyway
2) Scale down cached actors according to resource requirements

The main issue is when a trial completes and we decide to cache or not cache the actor, we don't know the next trial requirements, yet. We only know this after the next step() iteration when we created (or not created) a new trial.

Thus we eagerly cache actors and then check on the next step if we actually still need the actor. Note that this is not a full reconciliation (like PGs before) - we only _remove_ cached actors, never add.